### DISCUSSION TITLE
Lesson6: VirtualMachineError: revert

### DISCUSSION THREAD

**Discussion by ghost** [smartcontractkit/full-blockchain-solidity-course-py#843] [Q&A]
hello, i am gettting this error when i attempt to deploy fund_and_withdraw.py

```
Running 'scripts/fund_and_withdraw.py::main'...
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/fund_and_withdraw.py", line 13, in main
    fund()
  File "./scripts/fund_and_withdraw.py", line 8, in fund
    entrance_fee = fund_me.getEntranceFee()
  File "brownie/network/multicall.py", line 115, in _proxy_call
    result = ContractCall.__call__(*args, **kwargs)  # type: ignore
  File "brownie/network/contract.py", line 1729, in __call__
    return self.call(*args, block_identifier=block_identifier)
  File "brownie/network/contract.py", line 1533, in call
    raise VirtualMachineError(e) from None
VirtualMachineError: revert
Matts-MacBook-Air:brownie_fund_me mattwilson$
```

**Comment by ghost** [comment]
```Python
from brownie import FundMe
from scripts.helpful_scripts import get_account


def fund():
    fund_me = FundMe[-1]
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)


def main():
    fund()

```

**Reply by PatrickAlphaC** [reply]
What happens when you print out what fund_me is?

```python
from brownie import FundMe
from scripts.helpful_scripts import get_account


def fund():
    fund_me = FundMe[-1]
    print(fund_me)
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)


def main():
    fund()
```

**Comment by ghost** [comment]
```Python
from brownie import network, config, accounts, MockV3Aggregator
from web3 import Web3

DECIMALS = 8
STARTING_PRICE = 200000000000

LOCAL_BLOCKCHAIN_ENVIRONMENTS = ["development", "ganache-local3"]


def get_account():
    if network.show_active() in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        return accounts[0]
    else:
        return accounts.add(config["wallets"]["from_key"])


def deploy_mocks():
    print(f"The active network is {network.show_active()}")
    print("Deploying mocks...")
    if len(MockV3Aggregator) <= 0:
        MockV3Aggregator.deploy(
            DECIMALS, STARTING_PRICE, {
                "from": get_account()}
        )
    print("Mocks deployed!")

```

**Comment by ghost** [comment]
```Python
from brownie import network, config, FundMe, MockV3Aggregator

from scripts.helpful_scripts import get_account, deploy_mocks, MockV3Aggregator, LOCAL_BLOCKCHAIN_ENVIRONMENTS


def deploy_fund_me():
    account = get_account()
    if network.show_active() not in LOCAL_BLOCKCHAIN_ENVIRONMENTS:
        price_feed_address = config["networks"][network.show_active()][
            "eth_usd_price_feed"
        ]
    else:
        deploy_mocks()
        price_feed_address = MockV3Aggregator[-1].address

    fund_me = FundMe.deploy(
        {"from": account},
        publish_source=config["networks"][network.show_active()].get("verify"),
    )

    print(f"Contract deploy to {fund_me.address}")


def main():
    deploy_fund_me()

```

**Comment by ghost** [comment]
```Solidity
// SPDX-License-Identifier: MIT

// Smart contract that lets anyone deposit ETH into the contract
// Only the owner of the contract can withdraw the ETH
pragma solidity >=0.6.6 <0.9.0;

// Get the latest ETH/USD price from chainlink price feed
import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@chainlink/contracts/src/v0.6/vendor/SafeMathChainlink.sol";

contract FundMe {
    // safe math library check uint256 for integer overflows
    using SafeMathChainlink for uint256;

    //mapping to store which address depositeded how much ETH
    mapping(address => uint256) public addressToAmountFunded;
    // array of addresses who deposited
    address[] public funders;
    //address of the owner (who deployed the contract)
    address public owner;

    // the first person to deploy the contract is
    // the owner
    constructor() public {
        owner = msg.sender;
    }

    function fund() public payable {
        // 18 digit number to be compared with donated amount
        uint256 minimumUSD = 50 * 10**18;
        //is the donated amount less than 50USD?
        require(
            getConversionRate(msg.value) >= minimumUSD,
            "You need to spend more ETH!"
        );
        //if not, add to mapping and funders array
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }

    //function to get the version of the chainlink pricefeed
    function getVersion() public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(
            0x9326BFA02ADD2366b30bacB125260Af641031331
        );
        return priceFeed.version();
    }

    function getPrice() public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(
            0x9326BFA02ADD2366b30bacB125260Af641031331
        );
        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }

    // 1000000000
    function getConversionRate(uint256 ethAmount)
        public
        view
        returns (uint256)
    {
        uint256 ethPrice = getPrice();
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        // the actual ETH/USD conversation rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }

    function getEntranceFee() public view returns (uint256) {
        uint256 minimumUSD = 50 * 10**18;
        uint256 price = getPrice();
        uint256 precision = 1 * 10**18;
        return (minimumUSD * precision) / price;
    }

    //modifier: https://medium.com/coinmonks/solidity-tutorial-all-about-modifiers-a86cf81c14cb
    modifier onlyOwner() {
        //is the message sender owner of the contract?
        require(msg.sender == owner);

        _;
    }

    // onlyOwner modifer will first check the condition inside it
    // and
    // if true, withdraw function will be executed
    function withdraw() public payable onlyOwner {
        msg.sender.transfer(address(this).balance);

        //iterate through all the mappings and make them 0
        //since all the deposited amount has been withdrawn
        for (
            uint256 funderIndex = 0;
            funderIndex < funders.length;
            funderIndex++
        ) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        //funders array will be initialized to 0
        funders = new address[](0);
    }
}
```

**Comment by ghost** [comment]
```YAML
dotenv: .env
networks:
  kovan:
   eth_usd_price_feed: '0x9326BFA02ADD2366b30bacB125260Af641031331' 
   verify: True
  development:
   verify: False
  ganache-local3:
   verify: False

dependencies:
  # - <organization/repo>@<version>
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
compiler:
  solc:
    remappings:
      - '@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1'
wallets:
  from_key: ${PRIVATE_KEY}
```

**Comment by ghost** [comment]
i have tried to delete and remake the networks on ganache, double checked the .env and tried to restart vs code etc and nothing seems to work

**Comment by cromewar** [comment]
Hello @this is pretty common when brownie is not able to find the previous deployments of the contract so instead this:


```Python
def fund():
    fund_me = FundMe[-1]
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)
```
`
use fund_me = FundMe.deploy()` <- being this the script you use to deploy `deploy_fund_me()`.

**Reply by ghost** [reply]
```python
from brownie import FundMe, network, config
from scripts.helpful_scripts import get_account


def fund():
    account = get_account()
    fund_me = FundMe.deploy(
        {"from": account},
        publish_source=config["networks"][network.show_active()].get("verify"),
    )
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)


def main():
    fund()
```

**Reply by ghost** [reply]
i tried this and it gave the same error

**Reply by cromewar** [reply]
Meanwhile I do more research on this:

1. delete the build folder an try again.
2. change the verify to false.

**Reply by ghost** [reply]
hey i have tried these 2 and unfortunately neither have been the solution

**Comment by cromewar** [comment]
Doing further research I strongly think the error is here:

```Solidity
    function getPrice() public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(
            0x9326BFA02ADD2366b30bacB125260Af641031331
        );
        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }
```

answer must be marked as int. not int256, you'll transform it later.
Also make sure this: 0x9326BFA02ADD2366b30bacB125260Af641031331 price feed address correspond to the network you are working with.

**Reply by ghost** [reply]
when i try change it to int. answer and compile it automatically changes to int256 in vscode

**Reply by cromewar** [reply]
It shouldn't, jumm, did you checked the price_feed_address was correct?

**Reply by ghost** [reply]
yes i have checked this is a kovan price feed which is the one i've used for the course

**Reply by cromewar** [reply]
Jum, any ideas on this @PatrickAlphaC ?

**Reply by PatrickAlphaC** [reply]
What is the script you're running? (like the command in terminal) 

Additionally, could you run `node --version` and `ganache-cli --version`

**Reply by ghost** [reply]
Running 'scripts/fundandwithdraw.py::main'...
0x3352B5418E2F98D53A87AcCDCa991582FdFfaDBF
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/fundandwithdraw.py", line 14, in main
    fund()
  File "./scripts/fundandwithdraw.py", line 9, in fund
    entrance_fee = fund_me.getEntranceFee()
  File "brownie/network/multicall.py", line 115, in _proxy_call
    result = ContractCall.__call__(*args, **kwargs)  # type: ignore
  File "brownie/network/contract.py", line 1729, in __call__
    return self.call(*args, block_identifier=block_identifier)
  File "brownie/network/contract.py", line 1539, in call
    raise ValueError("No data was returned - the call likely reverted")
ValueError: No data was returned - the call likely reverted
Matts-MacBook-Air:brownie_fund_me mattwilson$ 

this is what happens when i print fund_me

**Reply by ghost** [reply]
v16.13.1 node version

**Reply by ghost** [reply]
and for some reason ganache-cli --version returns (command not found)

**Reply by PatrickAlphaC** [reply]
Ah. You'll need to install ganache-cli https://www.npmjs.com/package/ganache-cli

**Reply by cromewar** [reply]
Yeah that's it then, you must install ganache ;)

**Reply by ghost** [reply]
Matts-MacBook-Air:brownie_fund_me mattwilson$ npm install -g global-cli
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /usr/local/lib/node_modules/global-cli
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/global-cli'
npm ERR!  [Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/global-cli'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib/node_modules/global-cli'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mattwilson/.npm/_logs/2022-01-28T08_00_59_252Z-debug.log
Matts-MacBook-Air:brownie_fund_me mattwilson$

**Reply by ghost** [reply]
Matts-MacBook-Air:brownie_fund_me mattwilson$ yarn global add ganache-cli
yarn global v1.22.17
warning package.json: No license field
warning ../../package.json: No license field
[1/4] 🔍  Resolving packages...
warning ganache-cli@6.12.2: ganache-cli is now ganache; visit https://trfl.io/g7 for details
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
success Installed "ganache-cli@6.12.2" with binaries:
      - ganache-cli
✨  Done in 1.25s.
Matts-MacBook-Air:brownie_fund_me mattwilson$ ganache-cli --version
bash: ganache-cli: command not found

**Reply by ghost** [reply]
i do not understand why, but ganache-cli installs successfully and then when i lookup the version it continues to say not found

**Reply by cromewar** [reply]
OK there is the problem @

You are trying to install ganache globally using npm, your mac complains about it because installing global packages requires root privileges, so you could try uninstalling npm, and installing it trough nvm and brew. This will allow you to install global packages without root privileges.

In the other hand you can install it through Yarn, however as you can see it returns command not found. This means the Yarn packages are not added automatically to your path, so you must do it manually. Depending on your MacOS version the profile could be `.bashrc`, `zshrc` or `.bash_profile`. Look out how to add yarn packages to path on google ;).

**Reply by PatrickAlphaC** [reply]
or you can try `sudo yarn global add ganache-cli`
